streams/src/main/java/org/apache/kafka/streams/state/internals/StoreQueryUtils.java

```diff
@@ -351,6 +383,21 @@ public static <V> Function<byte[], V> getDeserializeValue(final StateSerdes<?, V
         return byteArray -> deserializer.deserialize(serdes.topic(), byteArray);
     }
 
+    public static <V> ValueIterator<VersionedRecord<V>> deserializeValueIterator(final StateSerdes<?, V> serdes,
+        final ValueIterator<VersionedRecord<byte[]>> rawValueIterator) {
+
+        final List<VersionedRecord<V>> versionedRecords = new ArrayList<>();
+        while (rawValueIterator.hasNext()) {
+            final VersionedRecord<byte[]> rawVersionedRecord = rawValueIterator.peek();
+            final Deserializer<V> valueDeserializer = serdes.valueDeserializer();
+            final long timestamp = rawVersionedRecord.timestamp();
+            final long validTo = rawVersionedRecord.validTo();
+            final V value = valueDeserializer.deserialize(serdes.topic(), rawVersionedRecord.value());
```

Comment:
I think we should not eagerly deserialize the result, but rather wrap the `rawValueIterator` with a `typesIterator` which does only deserialize the result that was requested via a `next()` call by the application.

Reply:
> I think we should not eagerly deserialize the result, but rather wrap the `rawValueIterator` with a `typesIterator` which does only deserialize the result that was requested via a `next()` call by the application.

Maybe we should discuss it more.

Reply:
Cf how `MeteredKeyValueStore` implements `RangeQuery`.

It takes the "raw" (byte[]) iterator, and wraps it: https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java#L276-L281

The "wrapper iterator" deserialized "on demand": https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/state/internals/MeteredKeyValueStore.java#L520-L523

This approach is better as it avoids expensive "ad hoc / upfront" deserialization cost (for example, a user might not even iterator over everything and close the iterator early, and even if they iterate fully, we avoid a "spike" and stretch out the deserialization overhead over a longer time frame what is preferable)

IMHO, we should try to do this no matter how we decide on the other open question (but it might make sense to hold off changing the code, until we figure a solution for the other open questions, too).

Reply:
> This approach is better as it avoids expensive "ad hoc / upfront" deserialization cost (for example, a user might not even iterator over everything and close the iterator early, and even if they iterate fully, we avoid a "spike" and stretch out the deserialization overhead over a longer time frame what is preferable)
> 
> IMHO, we should try to do this no matter how we decide on the other open question (but it might make sense to hold off changing the code, until we figure a solution for the other open questions, too).

Thanks Matthias. Yes, I got that. That's why, in my latest commit, I implemented lazy deserialization. Sorry, I just did not update my comment here.